### PR TITLE
Ternary search

### LINKED ISSUE #142: Ternary search
I wold like to make a ternary search implementation.

### PR DESCRIPTION
**Pull request by qrort**
Added ternary search implementation in C++.
Closes #142

### CODE REVIEW

**Comment by qrort** [comment]
Hello. Sorry for unopened issue, I hope you are not against any new implementation :)
I wonder why not the .cpp gets the build error, but the new point in algos list, Could you help me figure this out? Thanks/

**Comment by faheel** [comment]
> Sorry for unopened issue, I hope you are not against any new implementation :)

Not at all! 🙂  But it's always better to open an issue saying that you would like to contribute "algorithm X" or whatever and would be sending a PR for it, so that others know that someone is already working on "algorithm X". So please open an issue now for it.

> I wonder why not the .cpp gets the build error, but the new point in algos list

If you take a look at the Travis build log you'll see that it's the cpp file that causes the build to fail. The reason why the second commit is marked as "failed" is because [that's the way Travis works](https://blog.travis-ci.com/2017-03-22-introducing-auto-cancellation).

**Comment by qrort** [comment]
I am sorry, but I don't quite get it. What's the way I can figure out what's the problem in build on osx?

**Comment by faheel** [comment]
Nothing seems to be wrong on your part. Something's wrong with Travis not being able to install a Python library. I'll look into it and get back to you.

**Comment by qrort** [comment]
any upcoming info?

**Comment by qrort** [comment]
Btw is there a method to merge small commits into one so they don't look like now? Or do I need to cancel them and make one?

**Comment by qrort** [comment]
Now it is down to problems with Python again.

**Comment by faheel** [comment]
> Btw is there a method to merge small commits into one so they don't look like now?

Yes, read about [squashing commits in Git](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html).

**Comment by faheel** [comment]
I've made changes to the code that was causing the OSX builds to fail. Please update your branch with the new changes.

File: C++/test/searching/ternary_search.test.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+TEST_CASE("Integer cases", "[ternary search]") {
+    for (int i = 0; i < 10; i++) {
+        bool pattern = bool(nextRand24() % 2);
+        vector <int> a = nextRandUnimodal(nextRand24() % 1000, pattern);
+        int answer = solve(a, pattern);
+        REQUIRE(ternarySearch(a, pattern) == answer);
+    }
+}
```
Comment (faheel) [inline C++/test/searching/ternary_search.test.cpp:74]: You missed
```C++
#undef TERNARY_SEARCH_TEST
```

**Review by faheel — CHANGES_REQUESTED** [review]
Your indentation isn't uniform. Please change it to 4 spaces per level.

File: C++/test/searching/ternary_search.test.cpp (hunk 1 above)
Comment (qrort) [reply C++/test/searching/ternary_search.test.cpp:74]: done

**Review by faheel — APPROVED** [review]

**Comment by qrort** [comment]
Shall I close the request now?

**Comment by faheel** [comment]
No, I'll merge it in just a few minutes

**Comment by faheel** [comment]
It'll be easier for me if you squash your commits.

**Comment by faheel** [comment]
Thank you for your time and contribution @qrort! 😃

**Pull request merged**